src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.Unix.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using Microsoft.Win32.SafeHandles;
+
+namespace System.Formats.Tar
+{
+    // Unix specific methods for the TarEntry class.
+    public abstract partial class TarEntry
+    {
+        // Unix specific implementation of the method that extracts the current entry as a block device.
+        partial void ExtractAsBlockDevice(string destinationFileName)
+        {
+            Debug.Assert(EntryType is TarEntryType.BlockDevice);
+            Interop.CheckIo(Interop.Sys.CreateBlockDevice(destinationFileName, (int)Mode, (uint)_header._devMajor, (uint)_header._devMinor), destinationFileName);
+        }
+
+        // Unix specific implementation of the method that extracts the current entry as a character device.
+        partial void ExtractAsCharacterDevice(string destinationFileName)
+        {
+            Debug.Assert(EntryType is TarEntryType.CharacterDevice);
+            Interop.CheckIo(Interop.Sys.CreateCharacterDevice(destinationFileName, (int)Mode, (uint)_header._devMajor, (uint)_header._devMinor), destinationFileName);
+        }
+
+        // Unix specific implementation of the method that extracts the current entry as a fifo file.
+        partial void ExtractAsFifo(string destinationFileName)
+        {
+            Debug.Assert(EntryType is TarEntryType.Fifo);
+            Interop.CheckIo(Interop.Sys.MkFifo(destinationFileName, (int)Mode), destinationFileName);
+        }
+
+        // Unix specific implementation of the method that extracts the current entry as a hard link.
+        partial void ExtractAsHardLink(string destinationFileName)
+        {
+            Debug.Assert(EntryType is TarEntryType.HardLink);
+            Interop.CheckIo(Interop.Sys.Link(source: LinkName, link: destinationFileName), destinationFileName);
+        }
+
+        // Unix specific implementation of the method that specifies the file permissions of the extracted file.
+        partial void SetModeOnFile(SafeFileHandle handle, string destinationFileName)
+        {
+            // Only extract USR, GRP, and OTH file permissions, and ignore
+            // S_ISUID, S_ISGID, and S_ISVTX bits.
+            // It is off by default because it's possible that a file in an archive could have
+            // one of these bits set and, unknown to the person extracting, could allow others to
+            // execute the file as the user or group.
+            const int ExtractPermissionMask = 0x1FF;
+            int permissions = (int)Mode & ExtractPermissionMask;
+
+            // If the permissions weren't set at all, don't write the file's permissions.
+            if (permissions != 0)
+            {
+                Interop.CheckIo(Interop.Sys.FChMod(handle, permissions), destinationFileName);
```

Comment:
Unlike `tar` this doesn't respect `umask` while extracting.

Reply:
See my elaborate comment here: https://github.com/dotnet/runtime/pull/67883#issuecomment-1098829348.

I think we want to respect the umask since that is something that protects the user for unintentionally opening up permissions.

To do that, you can call the `internal` `SafeHandle.Open` on Unix that accepts `Interop.Sys.Permissions`:

https://github.com/dotnet/runtime/blob/dbf5c58a2e90a5dbe52248c5a4cad5e9aa6a4741/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs#L185-L190

Per that comment, at some point we'll also want to support applying all permissions (including sticky, setgroup, setuser), and setting ownership (uid, gid). The API will need to be extended to support that use-case.

cc @eerhardt